pallets/liquid-staking/src/lib.rs

```diff
@@ -1184,6 +1253,141 @@ pub mod pallet {
             Ok(())
         }
 
+        #[require_transactional]
+        fn do_multi_bond(
+            mut total_amount: BalanceOf<T>,
+            payee: RewardDestination<T::AccountId>,
+        ) -> DispatchResult {
+            let total_bonded = Self::get_total_bonded();
+            let new_total_bonded = total_bonded
+                .saturating_add(total_amount)
+                .min(Self::get_market_cap());
+            total_amount = new_total_bonded.saturating_sub(total_bonded);
+            if total_amount.is_zero() {
+                return Ok(());
+            }
+
+            let mut amounts: Vec<(DerivativeIndex, BalanceOf<T>)> = T::DerivativeIndexList::get()
+                .iter()
+                .map(|&index| (index, Self::bonded_of(index)))
+                .collect();
+            let new_avg_bonded = new_total_bonded.saturating_div(amounts.len() as BalanceOf<T>);
+
+            amounts.sort_by(|a, b| a.1.cmp(&b.1));
+
+            let mut distributions: Vec<(DerivativeIndex, BalanceOf<T>)> = vec![];
+            let mut remain = total_amount;
+
+            for (index, bonded) in amounts.into_iter() {
+                if remain.is_zero() {
+                    break;
+                }
+                let amount = Self::staking_ledger_cap()
+                    .saturating_sub(bonded)
+                    .min(new_avg_bonded.saturating_sub(bonded));
+                if amount.is_zero() || bonded.saturating_add(amount) < T::MinNominatorBond::get() {
+                    continue;
+                }
+                distributions.push((index, amount));
+                remain = remain.saturating_sub(amount);
+            }
+
+            for (index, amount) in distributions.into_iter() {
+                Self::do_bond(index, amount, payee.clone())?;
+            }
```

Comment:
Will keep looking deep into the algorithm,
and the `do_multi_unbond/do_multi_rebond/...`